### PR TITLE
feat(floating-label): Convert JS to TypeScript

### PR DESCRIPTION
**Pull request by acdvorak**
Refs #4225

### CODE REVIEW

File: packages/mdc-floating-label/adapter.ts [hunk 1]
```diff
@@ -21,50 +21,38 @@
...
-  getWidth() {}
+  getWidth(): number;
 
   /**
    * Registers an event listener on the root element for a given event.
-   * @param {string} evtType
-   * @param {function(!Event): undefined} handler
    */
-  registerInteractionHandler(evtType, handler) {}
+  registerInteractionHandler(evtType: string, handler: EventListener): void;
```
Comment (kfranqueiro) [inline packages/mdc-floating-label/adapter.ts]: Same comment as @moog16 left in https://github.com/material-components/material-components-web/pull/4373#discussion_r255133145

File: packages/mdc-floating-label/foundation.ts [hunk 2]
```diff
@@ -21,67 +21,59 @@
...
 
   /**
-   * Styles the label to produce the label shake for errors.
-   * @param {boolean} shouldShake adds shake class if true,
-   * otherwise removes shake class.
+   * Styles the label to produce a shake animation to indicate an error.
+   * @param shouldShake If true, adds the shake CSS class; otherwise, removes shake class.
    */
-  shake(shouldShake) {
+  shake(shouldShake: boolean): void {
```
Comment (kfranqueiro) [inline packages/mdc-floating-label/foundation.ts]: The `: void` returns are unnecessary, right?

Comment (moog16) [reply packages/mdc-floating-label/foundation.ts]: correct - I think they are inferred

File: packages/mdc-floating-label/adapter.ts (hunk 1 above)
Comment (acdvorak) [reply packages/mdc-floating-label/adapter.ts]: Done

File: packages/mdc-floating-label/foundation.ts (hunk 2 above)
Comment (acdvorak) [reply packages/mdc-floating-label/foundation.ts]: Removed

File: packages/mdc-floating-label/foundation.ts [hunk 3]
```diff
@@ -21,67 +21,59 @@
...
+  constructor(adapter: MDCFloatingLabelAdapter) {
     super(Object.assign(MDCFloatingLabelFoundation.defaultAdapter, adapter));
 
-    /** @private {function(!Event): undefined} */
     this.shakeAnimationEndHandler_ = () => this.handleShakeAnimationEnd_();
   }
 
   init() {
-    this.adapter_.registerInteractionHandler('animationend', this.shakeAnimationEndHandler_);
+    this.adapter_.registerInteractionHandler('animationend', this.shakeAnimationEndHandler_ as EventListener);
```
Comment (moog16) [inline packages/mdc-floating-label/foundation.ts]: since you changed the adapter, you should be able to remove the cast

Comment (acdvorak) [reply packages/mdc-floating-label/foundation.ts]: Done

**Pull request merged**